### PR TITLE
fix(torch): load weights only once

### PR DESCRIPTION
**Pull request by fantes**
this PR tries to cleanup model loading in torchlib.cc, previous version was extra cautious , leading to many unnecessary loadings which caused major slowdowns at inference

### CODE REVIEW

**Comment by beniz** [comment]
What happens when loading a model fails ? Loading is such a critical operation that I believe we may want to intercept exceptions of `torch::load`, log on a failed loading attempt, and rethrow so that it gets caught by DD upper levels, as per design.
What are your thoughts ?

File: src/backends/torch/torchgraphbackend.h [hunk 1]
```diff
@@ -103,27 +103,31 @@ namespace dd
     /**
      * \brief allocates torch modules base on real input dimension
      * @param inputdim input tensor dimension, in int64_t
+     * @return true if some allocation was done
      */
-    void finalize(std::vector<int64_t> inputdim);
+    bool finalize(std::vector<int64_t> inputdim);
```
Comment (BynaryCobweb) [inline src/backends/torch/torchgraphbackend.h:108]: Instead of returning a `bool`, there could be a flag (`need_reload` or something else) and a method `needs_reload()` to query this flag. This would avoid returning an unclear value (that needs to be documented) and make things more explicit in my opinion. And the graph could keep its "dirty" state in memory, so we do not need to keep track of it externally and it's more flexible to use (e.g. we do not need to save the result of the `finalize` method in a variable or a class field if we want to perform loading later in the code).

File: src/backends/torch/torchlib.cc [hunk 2]
```diff
@@ -622,16 +644,16 @@ namespace dd
...
 
+    // if (!this->_mlmodel._traced.empty())
+    //   this->_logger->info("Loading ml model from file {}.",
+    //                       this->_mlmodel._traced);
+    // if (!this->_mlmodel._proto.empty())
+    //   this->_logger->info("Loading ml model from file {}.",
+    //                       this->_mlmodel._proto);
+    // if (!this->_mlmodel._weights.empty())
+    //   this->_logger->info("Loading weights from file {}.",
+    //                       this->_mlmodel._weights);
```
Comment (BynaryCobweb) [inline src/backends/torch/torchlib.cc:655]: I think we can safely remove that

Comment (fantes) [reply src/backends/torch/torchlib.cc:655]: ok, done

**Comment by fantes** [comment]
> What happens when loading a model fails ? Loading is such a critical operation that I believe we may want to intercept exceptions of `torch::load`, log on a failed loading attempt, and rethrow so that it gets caught by DD upper levels, as per design.
> What are your thoughts ?

ok, done

File: src/backends/torch/torchgraphbackend.h (hunk 1 above)
Comment (fantes) [reply src/backends/torch/torchgraphbackend.h]: ok done

File: src/backends/torch/torchlib.cc [hunk 3]
```diff
@@ -175,17 +175,44 @@ namespace dd
...
+          _logger->info("net was reallocated due to input dim changes");
         // reload params after finalize
-        if (!tmodel._traced.empty())
-          torch::load(_graph, tmodel._traced, _device);
+        if (!tmodel._traced.empty() && _graph->needs_reload())
+          {
+            _logger->info("loading " + tmodel._traced);
+            try
+              {
+                torch::load(_graph, tmodel._traced, _device);
```
Comment (BynaryCobweb) [inline src/backends/torch/torchlib.cc:208]: Maybe add `set_need_reload(false)` or something like this?

File: src/backends/torch/torchgraphbackend.cc [hunk 4]
```diff
@@ -163,7 +163,12 @@ namespace dd
             std::get<1>(std::get<1>(full_output))); // last memory / c  value
         if (_lstm_continuation)
           {
-            _rnn_memories[opname_v] = std::get<1>(full_output);
+            if (_rnn_has_memories[opname_v])
+              _rnn_memories.erase(opname_v);
+            _rnn_memories.insert(
```
Comment (beniz) [inline src/backends/torch/torchgraphbackend.cc:168]: Just to make sure, this is not related to this PR, right ?

File: src/backends/torch/torchlib.cc [hunk 5]
```diff
@@ -175,17 +175,44 @@ namespace dd
...
+        if (!tmodel._native.empty())
+          {
+            _logger->info("loading " + tmodel._native);
+            try
+              {
+                torch::load(_native, tmodel._native, device);
+              }
+            catch (std::exception &e)
+              {
+                _logger->warn("unable to load " + tmodel._native);
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:189]: I'd use `_logger->error` instead.

File: src/backends/torch/torchlib.cc [hunk 6]
```diff
@@ -175,17 +175,44 @@ namespace dd
...
+        if (!tmodel._traced.empty() && _graph->needs_reload())
+          {
+            _logger->info("loading " + tmodel._traced);
+            try
+              {
+                torch::load(_graph, tmodel._traced, _device);
+              }
+            catch (std::exception &e)
+              {
+                _logger->warn("unable to load " + tmodel._traced);
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:212]: `_logger->error` ?

File: src/backends/torch/torchlib.cc [hunk 7]
```diff
@@ -175,17 +175,44 @@ namespace dd
...
         dims.insert(dims.begin(), 1); // dummy batch size
         _graph->finalize(dims);
+        if (_graph->needs_reload())
+          _logger->info("net was reallocated due to input dim changes");
         // reload params after finalize
-        if (!tmodel._traced.empty())
-          torch::load(_graph, tmodel._traced, _device);
+        if (!tmodel._traced.empty() && _graph->needs_reload())
+          {
+            _logger->info("loading " + tmodel._traced);
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:205]: I believe we may want a function for this, like `graph_model_load()` to make code lighter.

File: src/backends/torch/torchlib.cc [hunk 8]
```diff
@@ -399,13 +427,33 @@ namespace dd
   void TorchModule::load(TorchModel &model)
   {
     if (!model._traced.empty() && model._proto.empty())
-      _traced = std::make_shared<torch::jit::script::Module>(
-          torch::jit::load(model._traced, _device));
+      {
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:430]: Maybe put this block into a function, e.g. `traced_model_load()` ?

File: src/backends/torch/torchlib.cc [hunk 9]
```diff
@@ -884,7 +953,16 @@ namespace dd
...
         this->_logger->info("Restarting optimization from iter {}", it);
-        torch::load(*optimizer, this->_mlmodel._sstate);
+        this->_logger->info("loading " + this->_mlmodel._sstate);
+        try
+          {
+            torch::load(*optimizer, this->_mlmodel._sstate);
+          }
+        catch (std::exception &e)
+          {
+            this->_logger->warn("unable to load " + this->_mlmodel._sstate);
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:963]: `_logger->error` ?

File: src/backends/torch/torchlib.cc [hunk 10]
```diff
@@ -1379,7 +1457,7 @@ namespace dd
         unsupo.finalize(ad.getobj("parameters").getobj("output"), out,
                         static_cast<MLModel *>(&this->_mlmodel));
       }
-
+    empty_cuda_cache();
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:1460]: Might be unrelated to this PR ?

File: src/backends/torch/torchlib.h [hunk 11]
```diff
@@ -142,6 +142,8 @@ namespace dd
                                 the file where the weights are stored */
     unsigned int _nclasses = 0;
 
+    std::shared_ptr<spdlog::logger> _logger; /**< mllib logger. */
```
Comment (beniz) [inline src/backends/torch/torchlib.h:145]: This is already defined in `MLLib`, I don't think we should have it again here ?

File: src/caffegraphinput.cc [hunk 12]
```diff
@@ -189,7 +189,6 @@ namespace dd
         parse_simple_lstm(net);
         return 0;
       }
-    return 0;
+    return -1;
```
Comment (beniz) [inline src/caffegraphinput.cc:192]: This fix is unrelated to this PR.

File: src/mlservice.h [hunk 13]
```diff
@@ -158,7 +158,16 @@ namespace dd
       _init_parameters = ad.getobj("parameters");
       this->_inputc.init(_init_parameters.getobj("input"));
       this->_outputc.init(_init_parameters.getobj("output"));
-      this->init_mllib(_init_parameters.getobj("mllib"));
+      try
```
Comment (beniz) [inline src/mlservice.h:161]: Do you need this ? The main principle has always been to relay lower-level exceptions with their messages, and not obfuscate them at higher levels, when possible.

Comment (fantes) [reply src/mlservice.h:161]: I put this here because exception do not seem to be catched at upper level:  dede terminates in case of failure of init_mllib

File: src/backends/torch/torchgraphbackend.cc (hunk 4 above)
Comment (fantes) [reply src/backends/torch/torchgraphbackend.cc:168]: no, removed

File: src/backends/torch/torchlib.cc (hunk 5 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:189]: ok

File: src/backends/torch/torchlib.cc (hunk 6 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:212]: ok

File: src/backends/torch/torchlib.cc (hunk 7 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:205]: ok

File: src/backends/torch/torchlib.cc (hunk 8 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:430]: ok

File: src/backends/torch/torchlib.cc (hunk 9 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:963]: ok

File: src/backends/torch/torchlib.cc (hunk 10 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:1460]: ok; removed

File: src/backends/torch/torchlib.h (hunk 11 above)
Comment (fantes) [reply src/backends/torch/torchlib.h:145]: if one want to log directly from torchmodule, I do not see any other way to do

File: src/caffegraphinput.cc (hunk 12 above)
Comment (fantes) [reply src/caffegraphinput.cc:192]: this allows to throw smthg if proto do not contains a properly defined LSTM. Will change with smthg proper

File: src/backends/torch/torchlib.h (hunk 11 above)
Comment (beniz) [reply src/backends/torch/torchlib.h:145]: If torchmodule is uniquely associated to an instance of torchlib, I'd suggest that the logger from that torchlib object is used, no ?

File: src/mlservice.h (hunk 13 above)
Comment (beniz) [reply src/mlservice.h:161]: Let's make this another issue, all exceptions should be caught at upper levels.

File: src/backends/torch/torchlib.h (hunk 11 above)
Comment (fantes) [reply src/backends/torch/torchlib.h:145]: torchmodule does not have access to its container , so we give it the pointer to the logger so that it can use the same logger as its torchlib

File: src/mlservice.h (hunk 13 above)
Comment (fantes) [reply src/mlservice.h]: removed

File: src/backends/torch/torchlib.h (hunk 11 above)
Comment (beniz) [reply src/backends/torch/torchlib.h:145]: ok

Comment (beniz) [reply src/backends/torch/torchlib.h:145]: ok

File: src/backends/torch/torchlib.cc [hunk 14]
```diff
@@ -884,7 +954,16 @@ namespace dd
         size_t end = this->_mlmodel._sstate.rfind(".");
         it = std::stoi(this->_mlmodel._sstate.substr(start, end - start));
         this->_logger->info("Restarting optimization from iter {}", it);
-        torch::load(*optimizer, this->_mlmodel._sstate);
+        this->_logger->info("loading " + this->_mlmodel._sstate);
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:957]: Let's make this a function as well

File: src/backends/torch/torchlib.cc [hunk 15]
```diff
@@ -163,6 +163,84 @@ namespace dd
...
+          }
+      }
+  }
+
+  void TorchModule::classif_model_load(const TorchModel &model)
+  {
+    _logger->info("loading " + model._weights);
+    try
+      {
+        torch::load(_classif, model._weights, _device);
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:220]: This seems to me a generic loader, from `_model._weight`, I guess this is called `_classif` becauses it started by supporting classification ?

Additionnally, there's no support for the `device` parameter for that type of model ?

File: src/backends/torch/torchlib.cc [hunk 16]
```diff
@@ -163,6 +163,84 @@ namespace dd
...
+      }
+    catch (std::exception &e)
+      {
+        _logger->error("unable to load " + model._weights);
+        throw;
+      }
+  }
+
+  void TorchModule::traced_model_load(TorchModel &model)
```
Comment (beniz) [inline src/backends/torch/torchlib.cc:229]: Same remark here regarding the `device` parameter, if applicable.

File: src/backends/torch/torchlib.cc (hunk 15 above)
Comment (BynaryCobweb) [reply src/backends/torch/torchlib.cc:220]: _classif is a classification head added on top of traced model for classification tasks. If we want to support others heads, we may want to rename it `_head` and rename `model._weights` to `model._head_weights` ?

File: src/caffegraphinput.cc [hunk 17]
```diff
@@ -181,15 +183,15 @@ namespace dd
   {
     caffe::NetParameter net;
     if (!read_proto(filename, &net))
-      return -1;
+      throw MLLibBadParamException("unable to parse protofile");
```
Comment (BynaryCobweb) [inline src/caffegraphinput.cc:186]: In my understanding, this function both returns an error code *and* throws exceptions? Do we need to keep the error code, now it's always 0?

File: src/backends/torch/torchlib.cc (hunk 14 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:957]: ok

File: src/backends/torch/torchlib.cc (hunk 15 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:220]: about device : default behavior is to load to cpu, then `model.to(gpu)` , when i try to load directly to device, i sometimes have errors, so I will fall back to mandatory calls to `.to`

File: src/backends/torch/torchlib.cc (hunk 16 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:229]: see above : I prefer to remove direct loading to device, there is something that does not work with it.

File: src/caffegraphinput.cc (hunk 17 above)
Comment (fantes) [reply src/caffegraphinput.cc:186]: ok, removed error code

File: src/backends/torch/torchlib.cc (hunk 15 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc:220]: about _model_weight : indeed there will need some cleanup at some time: we can either load a traced model, some weight to put in _classif head, a prototxt, weight to a model that is defined by a prototxt (and that use the same extension as traced nets), and weight for a native model. We should clean up file extensions in order to be able to load several of them in the future.

Comment (beniz) [reply src/backends/torch/torchlib.cc:219]: > about device : default behavior is to load to cpu, then `model.to(gpu)` , when i try to load directly to device, i sometimes have errors, so I will fall back to mandatory calls to `.to`

Can we double check on that ? Like, are we loading the model twice effectively ? I don't want to slow this PR down, but I fee if the PR is called 'load weights only once', let's check on that :)

Comment (fantes) [reply src/backends/torch/torchlib.cc:219]: check on what ? do you want to try to load models directly to device ?

**Review by beniz — APPROVED** [review]

File: src/backends/torch/torchlib.cc [hunk 18]
```diff
@@ -362,6 +442,7 @@ namespace dd
 
     if (!_classif_layer_file.empty())
       {
+        _logger->info("loading " + _classif_layer_file);
         torch::load(_classif, _classif_layer_file, _device);
```
Comment (BynaryCobweb) [inline src/backends/torch/torchlib.cc:446]: Model is loaded here as well (This method is called during training to finalize the classification layer with some example data)
It would be great to replace torch::load with classif_model_load

File: src/backends/torch/torchlib.cc (hunk 3 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc]: done

File: src/backends/torch/torchlib.cc (hunk 18 above)
Comment (fantes) [reply src/backends/torch/torchlib.cc]: ok done
